### DISCUSSION TITLE
Requesting: "Reservation System" Booking

### DISCUSSION THREAD

**Discussion by DrMattBrown** [btcpayserver/btcpayserver#884] [Plugins & integrations]
Would it be possible to create a feature that lets someone schedule an appointment whether that be at a salon, yoga studio, doctor's office, restaurant or hotel? Functionality would be like just about any booking system out there. 

If I was to suggest anything to emulate it would be the bookly plugin for WordPress.. why? just because it seems cobbled together in a good way and they have good documentation: https://support.booking-wp-plugin.com/hc/en-us/categories/200390551-Documentation

The very basic level of functionality here could be done in many ways, but I woudl suggest looking into creating a smart contract or a Ln payment channel. If smart contract it woudl just need to hold the funds in an escrow address for a specified amount of time prior to the reserved service. That would allow the buyer to cancel or reschedule without penalty. Once X time has passed, the payment is officially sent to the seller and the buyer can no longer cancel.

**Comment by pavlenex** [comment]
Speaking of WordPress, if there's a booking plugin for WooCommerce, it should work with BTCPay for sure. Something like [this](https://woocommerce.com/products/woocommerce-bookings/) or any of [these](https://www.collectiveray.com/wp/plugins/woocommerce-appointments.html) should work.

**Comment by NicolasDorier** [comment]
@DrMattBrown I think this functionality is WAY better to be handled by wordpress. 

Because you don't want only booking system, you also want a way to design your website and all.

**Comment by DrMattBrown** [comment]
I do not disagree that using wordpress would more easily allow for all the customizations etc that the above wordpress plugin has, but I do feel that a basic "delayed settlement" or "Reservation" where someone elects to purchase a service and they can select when they would like to schedule that service for ( just select a date and time on a calendar, or even a drop down to start) and the transaction is finalized within a specified amount of time from that selected time ( say 24hours before). 

This delayed settlement could be used for other use cases in the future. 

The fancy bells and whistles of the bookly system are not needed really, but I was just trying to find something with robust documentation.

**Comment by pavlenex** [comment]
You can most likely do this with [Payment Requests](https://docs.btcpayserver.org/deployment/manualdeployment) we have, though there's no service schedule.
Possibly with payment forms #429 as well

**Comment by NicolasDorier** [comment]
@pavlenex I think the only thing needed is the wordpress booking + btcpay woocommerce plugin => Done.

**Comment by DrMattBrown** [comment]
@NicolasDorier Yes again. All this can easily be done with wordpress and woocommerce. 

I was only adding the request as I understood there is a push to let BTCpayserver become it's own stand-alone solution that didn't need to rely on a separate website or web server; running it all on a rasberry pi and routing through TOR.

**Comment by pavlenex** [comment]
Maybe we can just tweak what we already have to be able to do this, just like add a checkbox or something.

So it should look like a payment request, but with a calendar where customer can select the date before they pay, then that date should be parsed somewhere in BTCPay invoice so that a merchant can see when the customer scheduled it.

I think payment requests would be the perfect solution, but there is no "Scheduled date" in our invoice which again depends on my "BTCPay Forms" issue. So maybe we can just add schedule form in BTCPay Forms, and that would be better as it would allow more customization.

**Comment by DrMattBrown** [comment]
I think that sounds like an ideal implementation strategy... again, all the extra confirmations and tracking the appointment are all gravy really. All this is is a scheduled delayed payment.

**Comment by pavlenex** [comment]
Oh, there's no way we can do a scheduled delayed payment. How would that work with volatility? 
Or you want the customer to book, then get an invoice via email at the time he scheduled?

**Comment by DrMattBrown** [comment]
Hmm.. good point.. The goal is to secure a payment for a service at the time of booking to help reduce the likelihood of a no-show. Perhaps for now volatility on the main chain could make this difficult. I do suppose you could do something with a lightning payment channel? Like open a channel and fund it with more than enough for the visit at time of booking and then execute the transaction just before, but still many wouldn't want to lock their coins in a lightning channel like this. I'm not sure what the ideal fix might be for the volatility issue.

**Comment by pavlenex** [comment]
Not the volatility on the main chain, but the volatility of the BTC/FIAT, in general, would make this very difficult.

What you're basically asking is some sort of `subscription` #461 where the customer would get an email with a link to a payment request.

After thinking, I think this is way better handled with WordPress as I originally said. There are just way too many missing pieces in BTCPay to currently handle it.

Basically :
- We need to add a scheduling option in payment requests
- We need to add `scheduled time` field in the invoice
- We need to add #461 

By the way, WordPress site can be [hosted together on your BTCPay](https://docs.btcpayserver.org/integrations/woocommerce#deploying-woocommerce-from-btcpay-server), so you can host them on the same machine.

**Comment by DrMattBrown** [comment]
Seems like a reasonable approach. 

I know I am not using the correct terminology here. I do understand that the volatility is between fiat/BTC and not anything with BTC itself. 

This is very similar to what a guy named David Zimbeck coded like 5 years ago for BTC with hopes of using it for auctions or selling goods online, but again, the fiat volatility thing was an issue. 

Regardless, please feel free to close this request and/or merge with #461 as it does sound like ultimately that may be the path forward with this issue. 

Again, the only reason for this request is to come up with a solution to help BTCpayserver do all of the things.  Until then, wordpress/woocommerce are great tools and make the entire process WAY easier.

**Comment by pavlenex** [comment]
I know @justinmoon had a similar request, I'll leave it, perhaps someone wants to add a simple booking system.
